### PR TITLE
add cache in events to avoid duplication

### PR DESCRIPTION
**Pull request by rahulsawra98**
This PR adds a cache to eventRecorder that stores a unique key generated and event as a value.

### CODE REVIEW

File: pkg/events/events.go [hunk 1]
```diff
@@ -194,10 +221,26 @@ func (er *eventRecorder) RecordEvent(ctx context.Context, e *Event) error {
...
-		er.Logger.With("error", err, "event", ev).Error("Unable to create event")
+	// Check if there is already an event of the same type in the cache
+	key := getEventKey(ev)
+	er.cacheLock.Lock()
+	lastSeenEvent,ok := er.cache.Get(key)
+	er.cacheLock.Unlock()
+	if !ok{
+		ev.FirstTimestamp = t
+		if err := er.Client.Create(ctx, ev); err != nil {
+			er.Logger.With("error", err, "event", ev).Error("Unable to create event")
```
Comment (kmjayadeep) [inline pkg/events/events.go:232]: error should be returned back from the function

File: pkg/events/events.go [hunk 2]
```diff
@@ -194,10 +221,26 @@ func (er *eventRecorder) RecordEvent(ctx context.Context, e *Event) error {
...
+		} else {
+			er.cache.Add(key,ev)
+		}
+	} else {
+		// event already present in cache
+		e := lastSeenEvent.(*corev1.Event)
+		e.Count++
+		e.LastTimestamp = t
+		if err := er.Client.Update(ctx, e); err != nil {
+			er.Logger.With("error", err, "event", ev).Error("Unable to update event")
```
Comment (kmjayadeep) [inline pkg/events/events.go:242]: error should be returned back from the function

**Review by kmjayadeep — CHANGES_REQUESTED** [review]
The approach looks good to me. I would suggest to add some test cases as well, since there might be some corner cases. 

I would propose to cover at least the following cases

- Creating event for the first time - make sure cache is updated
- Creating multiple events with mix and match of same and different attributes. Make sure that getEventKey is working as expected
- Invalidation of cache after a certain time period
- Validate event count

I think a table driven test would be the best approach here since there are so many cases to cover

**Comment by rahulsawra98** [comment]
> The approach looks good to me. I would suggest to add some test cases as well, since there might be some corner cases.
> 
> I would propose to cover at least the following cases
> 
> * Creating event for the first time - make sure cache is updated
> * Creating multiple events with mix and match of same and different attributes. Make sure that getEventKey is working as expected
> * Invalidation of cache after a certain time period
> * Validate event count
> 
> I think a table driven test would be the best approach here since there are so many cases to cover

Done! PTAL!

**Review by PuranAdhikari — APPROVED** [review]

File: pkg/events/events.go [hunk 3]
```diff
@@ -81,12 +87,34 @@ type Event struct {
...
+		event.InvolvedObject.FieldPath,
+		string(event.InvolvedObject.UID),
+		event.InvolvedObject.APIVersion,
+		event.Type,
+		event.Reason,
+		event.Message,
+		event.Labels["sliceName"],
+		event.Labels["sliceCluster"],
+	},
+		"")
```
Comment (narmidm) [inline pkg/events/events.go:107]: please add - the project & name

**Review by narmidm — APPROVED** [review]
